### PR TITLE
Add monthly P3A metrics for Ethereum signatures

### PR DESCRIPTION
**Pull request by DJAndries**
Resolves brave/brave-browser#28572

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Will write soon.

### CODE REVIEW

**Review by kdenhartog — APPROVED** [review]
LGTM - thanks for the quick addition

File: components/brave_wallet/browser/ethereum_provider_impl.h [hunk 1]
```diff
@@ -257,6 +258,7 @@ class EthereumProviderImpl final
                            const absl::optional<std::string>& domain_hash,
                            const absl::optional<std::string>& primary_hash,
                            bool is_eip712,
+                           bool is_eth_sign,
```
Comment (darkdh) [inline components/brave_wallet/browser/ethereum_provider_impl.h:261]: Please do not add `is_eth_sign` when `is_eip712` is false already means doing `eth_sign`

**Review by darkdh — CHANGES_REQUESTED** [review]

File: components/brave_wallet/browser/ethereum_provider_impl.h [hunk 2]
```diff
@@ -80,7 +80,8 @@ class EthereumProviderImpl final
   void SignMessage(const std::string& address,
                    const std::string& message,
                    RequestCallback callback,
-                   base::Value id);
+                   base::Value id,
+                   bool is_eth_sign);
```
Comment (darkdh) [inline components/brave_wallet/browser/ethereum_provider_impl.h:84]: And also don't do this, `SignMessage` is dedicated for `eth_sign` and `SignTypedMessage` for eip712

Comment (kdenhartog) [reply components/brave_wallet/browser/ethereum_provider_impl.h:84]: It looks like `SignMessage` is covering both the `eth_sign` and `personal_sign` APIs. In the ideal case, we'd still want to keep `personal_sign` and deprecate  `eth_sign` only.

The code on [1142 to 1156 in the implementation](https://github.com/brave/brave-core/pull/17266/files/e7101bfec929e64a72dfd809dbd5fe1a439e691e..2f5c8d267b7190d75f91ea919cfd3fe0e4b69119#diff-5b97dccf71a954164c80d8d4453cc9b47cf0c58ae4c771a849187dc5bf80f888R1142-R1156) is what gave me that idea

Comment (DJAndries) [reply components/brave_wallet/browser/ethereum_provider_impl.h:84]: ^ this. Originally I had planned to use `is_eip712 == false` as the condition for triggering the P3A event, but since EIP-191 includes both `eth_sign` and `personal_sign`, I felt like I had no choice.

@darkdh alternative suggestions are welcome, thanks

File: components/brave_wallet/browser/ethereum_provider_impl.h (hunk 1 above)
Comment (DJAndries) [reply components/brave_wallet/browser/ethereum_provider_impl.h:261]: see below

File: components/brave_wallet/browser/ethereum_provider_impl.h (hunk 2 above)
Comment (darkdh) [reply components/brave_wallet/browser/ethereum_provider_impl.h:84]: If we are aiming to deprecate `eth_sign` then this makes sense.

**Review by darkdh — APPROVED** [review]

**Comment by kdenhartog** [comment]
For context if someone comes to take a look at this PR later, this P3A question is being added to get some rough understanding about how often `eth_sign` is being used. We want to see what will be necessary to start pushing for deprecation of this RPC method across the ecosystem. Today, there are many attacks conducted because this API doesn't display the full message of what's being signed to the user and this creates a lot of harm such as wallet drain attacks and phishing scams. The intent is to remove this P3A question once we've got a rough understanding of the amount of usage this API sees in about 3 months time.

File: components/p3a_utils/feature_usage.cc [hunk 3]
```diff
@@ -15,7 +15,7 @@ namespace p3a_utils {
 
 namespace {
 
-constexpr int kDaysInMonthBuckets[] = {0, 1, 2, 5, 10, 15, 20, 100};
+constexpr uint64_t kDaysInMonthBuckets[] = {0, 1, 2, 5, 10, 15, 20, 100};
```
Comment (iefremov) [inline components/p3a_utils/feature_usage.cc:18]: i wonder why this change is needed? why not use 32-bit unsigned at least?

**Review by iefremov — APPROVED** [review]

**Comment by kdenhartog** [comment]
I'm good with closing this one without merging and instead going the route of metamask where we just turn eth_sign off by default or remove it.
